File: Pinball App 2/Pinball App 2/practice/PracticeJournalSummaryStyling.swift
import SwiftUI

private enum PracticeJournalSummaryTokenColor {
    case primary
    case game
    case screen
    case score
    case note
}

private struct PracticeJournalSummaryToken {
    let text: String
    let color: PracticeJournalSummaryTokenColor
}

func styledPracticeJournalSummary(_ summary: String) -> Text {
    let tokens = practiceJournalSummaryTokens(summary)
    return tokens.reduce(Text("")) { partial, token in
        let segment = Text(token.text)
            .foregroundStyle(colorForPracticeJournalToken(token.color))
        switch token.color {
        case .primary:
            return partial + segment
        case .game, .screen, .score:
            return partial + segment.bold()
        case .note:
            return partial + segment
        }
    }
}

private func colorForPracticeJournalToken(_ token: PracticeJournalSummaryTokenColor) -> Color {
    let game = Color(UIColor { trait in
        trait.userInterfaceStyle == .dark ? UIColor(red: 0.48, green: 0.93, blue: 0.96, alpha: 1) : UIColor(red: 0.04, green: 0.43, blue: 0.38, alpha: 1)
    })
    let screen = Color(UIColor { trait in
        trait.userInterfaceStyle == .dark ? UIColor(red: 0.69, green: 0.82, blue: 0.99, alpha: 1) : UIColor(red: 0.10, green: 0.26, blue: 0.78, alpha: 1)
    })
    let score = Color(UIColor { trait in
        trait.userInterfaceStyle == .dark ? UIColor(red: 0.99, green: 0.82, blue: 0.24, alpha: 1) : UIColor(red: 0.61, green: 0.29, blue: 0.03, alpha: 1)
    })
    let note = Color(UIColor { trait in
        trait.userInterfaceStyle == .dark ? UIColor(red: 0.90, green: 0.93, blue: 0.98, alpha: 1) : UIColor(red: 0.23, green: 0.28, blue: 0.37, alpha: 1)
    })
    switch token {
    case .primary: return .primary
    case .game: return game
    case .screen: return screen
    case .score: return score
    case .note: return note
    }
}

private func practiceJournalSummaryTokens(_ summary: String) -> [PracticeJournalSummaryToken] {
    if let score = parseScoreSummary(summary) {
        return [
            .init(text: "Score: ", color: .primary),
            .init(text: score.value, color: .score),
            .init(text: " • ", color: .primary),
            .init(text: score.game, color: .game),
            .init(text: " (\(score.context))", color: .screen),
        ]
    }

    if let practice = parseStructuredPracticeSummary(summary) {
        return practice
    }

    if let study = parseStructuredStudySummary(summary) {
        return study
    }

    if let gameNote = parseStructuredGameNoteSummary(summary) {
        return gameNote
    }

    if let bullet = parseBulletGameSummary(summary) {
        return [
            .init(text: bullet.prefix, color: .primary),
            .init(text: "\n• ", color: .screen),
            .init(text: bullet.game, color: .game),
        ]
    }

    if let lib = parseLibrarySummary(summary) {
        return lib
    }

    if let practicePlayfield = parsePracticePlayfieldSummary(summary) {
        return practicePlayfield
    }
    if let rulesheet = parsePracticeRulesheetSummary(summary) {
        return rulesheet
    }
    if let video = parsePracticeVideoSummary(summary) {
        return video
    }
    if let progress = parsePracticeProgressSummary(summary) {
        return progress
    }
    if let browsed = parsePracticeBrowsedSummary(summary) {
        return browsed
    }

    return [.init(text: summary, color: .primary)]
}

private func parseScoreSummary(_ summary: String) -> (value: String, game: String, context: String)? {
    guard summary.hasPrefix("Score: ") else { return nil }
    let rest = String(summary.dropFirst("Score: ".count))
    guard let bulletRange = rest.range(of: " • "),
          let contextStart = rest.lastIndex(of: "("),
          rest.hasSuffix(")"),
          contextStart > bulletRange.upperBound else { return nil }
    let value = String(rest[..<bulletRange.lowerBound])
    let gamePart = String(rest[bulletRange.upperBound..<contextStart]).trimmingCharacters(in: .whitespaces)
    let context = String(rest[rest.index(after: contextStart)..<rest.index(before: rest.endIndex)])
    guard !value.isEmpty, !gamePart.isEmpty, !context.isEmpty else { return nil }
    return (value, gamePart, context)
}

private func parseBulletGameSummary(_ summary: String) -> (prefix: String, game: String)? {
    guard let range = summary.range(of: "\n• ") else { return nil }
    let prefix = String(summary[..<range.lowerBound])
    let game = String(summary[range.upperBound...])
    guard !game.isEmpty else { return nil }
    return (prefix, game)
}

private func parseStructuredPracticeSummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    guard summary.hasPrefix("Practice:\n"), let range = summary.range(of: "\n• ") else { return nil }
    let body = String(summary[summary.index(summary.startIndex, offsetBy: "Practice:\n".count)..<range.lowerBound])
    let game = String(summary[range.upperBound...])
    guard !game.isEmpty else { return nil }

    var lines = body.components(separatedBy: "\n")
    while lines.last?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == true { lines.removeLast() }
    let firstLine = lines.first?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    let valueLine = firstLine.isEmpty ? "Practice session" : firstLine
    let noteText = lines.dropFirst().joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)

    var tokens: [PracticeJournalSummaryToken] = [
        .init(text: "Practice", color: .screen),
        .init(text: ":\n", color: .primary),
        .init(text: valueLine, color: .screen),
    ]
    if !noteText.isEmpty {
        tokens.append(.init(text: "\n", color: .primary))
        tokens.append(.init(text: noteText, color: .note))
    }
    tokens.append(.init(text: "\n• ", color: .screen))
    tokens.append(.init(text: game, color: .game))
    return tokens
}

private func parseStructuredGameNoteSummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    guard summary.hasPrefix("Game Note:\n"), let range = summary.range(of: "\n• ") else { return nil }
    let note = String(summary[summary.index(summary.startIndex, offsetBy: "Game Note:\n".count)..<range.lowerBound])
    let game = String(summary[range.upperBound...])
    guard !game.isEmpty else { return nil }
    return [
        .init(text: "Game Note", color: .screen),
        .init(text: ":\n", color: .primary),
        .init(text: note, color: .note),
        .init(text: "\n• ", color: .screen),
        .init(text: game, color: .game),
    ]
}

private func parseStructuredStudySummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    let headers: [(label: String, titleToken: String)] = [
        ("Rulesheet", "Rulesheet"),
        ("Tutorial Video", "Tutorial Video"),
        ("Gameplay Video", "Gameplay Video"),
        ("Playfield", "Playfield")
    ]
    guard let header = headers.first(where: { summary.hasPrefix($0.label + ":\n") }),
          let range = summary.range(of: "\n• ")
    else { return nil }

    let body = String(summary[summary.index(summary.startIndex, offsetBy: header.label.count + 2)..<range.lowerBound])
    let game = String(summary[range.upperBound...])
    guard !game.isEmpty else { return nil }
    var lines = body.components(separatedBy: "\n")
    while lines.last?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == true { lines.removeLast() }
    let valueLine = lines.first?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    let noteText = lines.dropFirst().joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
    guard !valueLine.isEmpty else { return nil }

    var tokens: [PracticeJournalSummaryToken] = [
        .init(text: header.titleToken, color: .screen),
        .init(text: ":\n", color: .primary)
    ]
    if let progress = valueLine.stripProgressPrefix() {
        tokens.append(.init(text: "Progress", color: .screen))
        tokens.append(.init(text: ": ", color: .primary))
        tokens.append(.init(text: progress, color: .screen))
    } else if valueLine.caseInsensitiveCompare("Viewed playfield") == .orderedSame {
        tokens.append(.init(text: "Viewed ", color: .primary))
        tokens.append(.init(text: "playfield", color: .screen))
    } else {
        tokens.append(.init(text: valueLine, color: .screen))
    }
    if !noteText.isEmpty {
        tokens.append(.init(text: "\n", color: .primary))
        tokens.append(.init(text: noteText, color: .note))
    }
    tokens.append(.init(text: "\n• ", color: .screen))
    tokens.append(.init(text: game, color: .game))
    return tokens
}

private func parseLibrarySummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    if summary.hasPrefix("Browsed "), summary.hasSuffix(" in Library") {
        let game = String(summary.dropFirst("Browsed ".count).dropLast(" in Library".count))
        return [
            .init(text: "Browsed ", color: .primary),
            .init(text: game, color: .game),
            .init(text: " in ", color: .primary),
            .init(text: "Library", color: .screen),
        ]
    }
    if summary.hasPrefix("Opened "), summary.hasSuffix(" rulesheet from Library") {
        let game = String(summary.dropFirst("Opened ".count).dropLast(" rulesheet from Library".count))
        return [
            .init(text: "Opened ", color: .primary),
            .init(text: game, color: .game),
            .init(text: " ", color: .primary),
            .init(text: "rulesheet", color: .screen),
            .init(text: " from ", color: .primary),
            .init(text: "Library", color: .screen),
        ]
    }
    if summary.hasPrefix("Opened "), summary.hasSuffix(" playfield image from Library") {
        let game = String(summary.dropFirst("Opened ".count).dropLast(" playfield image from Library".count))
        return [
            .init(text: "Opened ", color: .primary),
            .init(text: game, color: .game),
            .init(text: " ", color: .primary),
            .init(text: "playfield image", color: .screen),
            .init(text: " from ", color: .primary),
            .init(text: "Library", color: .screen),
        ]
    }
    if summary.hasPrefix("Opened "), summary.hasSuffix(" in Library"), let videoRange = summary.range(of: " video for ") {
        let detail = String(summary[summary.index(summary.startIndex, offsetBy: "Opened ".count)..<videoRange.lowerBound])
        let game = String(summary[videoRange.upperBound..<summary.index(summary.endIndex, offsetBy: -" in Library".count)])
        return [
            .init(text: "Opened ", color: .primary),
            .init(text: detail, color: .screen),
            .init(text: " ", color: .primary),
            .init(text: "video", color: .screen),
            .init(text: " for ", color: .primary),
            .init(text: game, color: .game),
            .init(text: " in ", color: .primary),
            .init(text: "Library", color: .screen),
        ]
    }
    return nil
}

private extension String {
    func stripProgressPrefix() -> String? {
        guard hasPrefix("Progress: ") else { return nil }
        let out = String(dropFirst("Progress: ".count))
        return out.isEmpty ? nil : out
    }
}

private func parsePracticePlayfieldSummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    guard summary.hasPrefix("Viewed "),
          let range = summary.range(of: " playfield")
    else { return nil }

    let game = String(summary[summary.index(summary.startIndex, offsetBy: "Viewed ".count)..<range.lowerBound])
    guard !game.isEmpty else { return nil }

    let suffix = String(summary[range.upperBound...])
    let note: String?
    if suffix.hasPrefix(": ") {
        note = String(suffix.dropFirst(2))
    } else if suffix.isEmpty {
        note = nil
    } else {
        return nil
    }

    var tokens: [PracticeJournalSummaryToken] = [
        .init(text: "Viewed ", color: .primary),
        .init(text: game, color: .game),
        .init(text: " ", color: .primary),
        .init(text: "playfield", color: .screen),
    ]
    if let note, !note.isEmpty {
        tokens.append(.init(text: ": \(note)", color: .primary))
    }
    return tokens
}

private func parsePracticeRulesheetSummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    guard summary.hasPrefix("Read "), summary.hasSuffix(" rulesheet") else { return nil }
    let body = String(summary.dropFirst("Read ".count).dropLast(" rulesheet".count))
    if let ofRange = body.range(of: " of ") {
        let progress = String(body[..<ofRange.lowerBound])
        let game = String(body[ofRange.upperBound...])
        guard !progress.isEmpty, !game.isEmpty else { return nil }
        return [
            .init(text: "Read ", color: .primary),
            .init(text: progress, color: .screen),
            .init(text: " of ", color: .primary),
            .init(text: game, color: .game),
            .init(text: " ", color: .primary),
            .init(text: "rulesheet", color: .screen),
        ]
    }
    guard !body.isEmpty else { return nil }
    return [
        .init(text: "Read ", color: .primary),
        .init(text: body, color: .game),
        .init(text: " ", color: .primary),
        .init(text: "rulesheet", color: .screen),
    ]
}

private func parsePracticeVideoSummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    for prefix in ["Tutorial for ", "Gameplay for "] {
        guard summary.hasPrefix(prefix), let range = summary.range(of: ": ") else { continue }
        let kind = prefix.hasPrefix("Tutorial") ? "Tutorial" : "Gameplay"
        let game = String(summary[summary.index(summary.startIndex, offsetBy: prefix.count)..<range.lowerBound])
        let value = String(summary[range.upperBound...])
        guard !game.isEmpty else { return nil }
        return [
            .init(text: kind, color: .screen),
            .init(text: " for ", color: .primary),
            .init(text: game, color: .game),
            .init(text: ": ", color: .primary),
            .init(text: value, color: .primary),
        ]
    }

    if summary.hasPrefix("Updated tutorial progress for ") {
        let game = String(summary.dropFirst("Updated tutorial progress for ".count))
        return [
            .init(text: "Updated ", color: .primary),
            .init(text: "tutorial", color: .screen),
            .init(text: " progress for ", color: .primary),
            .init(text: game, color: .game),
        ]
    }
    if summary.hasPrefix("Updated gameplay progress for ") {
        let game = String(summary.dropFirst("Updated gameplay progress for ".count))
        return [
            .init(text: "Updated ", color: .primary),
            .init(text: "gameplay", color: .screen),
            .init(text: " progress for ", color: .primary),
            .init(text: game, color: .game),
        ]
    }
    return nil
}

private func parsePracticeProgressSummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    if summary.hasPrefix("Practice progress "), let onRange = summary.range(of: " on ") {
        let value = String(summary[summary.index(summary.startIndex, offsetBy: "Practice progress ".count)..<onRange.lowerBound])
        let game = String(summary[onRange.upperBound...])
        guard !value.isEmpty, !game.isEmpty else { return nil }
        return [
            .init(text: "Practice", color: .screen),
            .init(text: " progress ", color: .primary),
            .init(text: value, color: .screen),
            .init(text: " on ", color: .primary),
            .init(text: game, color: .game),
        ]
    }
    if summary.hasPrefix("Logged practice for ") {
        let game = String(summary.dropFirst("Logged practice for ".count))
        guard !game.isEmpty else { return nil }
        return [
            .init(text: "Logged ", color: .primary),
            .init(text: "practice", color: .screen),
            .init(text: " for ", color: .primary),
            .init(text: game, color: .game),
        ]
    }
    return nil
}

private func parsePracticeBrowsedSummary(_ summary: String) -> [PracticeJournalSummaryToken]? {
    guard summary.hasPrefix("Browsed "), !summary.hasSuffix(" in Library") else { return nil }
    let game = String(summary.dropFirst("Browsed ".count))
    guard !game.isEmpty else { return nil }
    return [
        .init(text: "Browsed ", color: .primary),
        .init(text: game, color: .game),
    ]
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinballandroid/practice/PracticeJournalSummaryStyling.kt
package com.pillyliu.pinballandroid.practice

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.withStyle

private enum class JournalTokenColor { Primary, Game, Screen, Score, Note }

private data class JournalToken(val text: String, val color: JournalTokenColor)

@Composable
internal fun StyledPracticeJournalSummaryText(
    summary: String,
    style: TextStyle,
    modifier: Modifier = Modifier,
) {
    val colors = MaterialTheme.colorScheme
    val isDark = isSystemInDarkTheme()
    val annotated = buildAnnotatedString {
        journalSummaryTokens(summary).forEach { token ->
            withStyle(
                SpanStyle(
                    color = when (token.color) {
                        JournalTokenColor.Primary -> colors.onSurface
                        JournalTokenColor.Game -> if (isDark) androidx.compose.ui.graphics.Color(0xFF6EE7F9) else androidx.compose.ui.graphics.Color(0xFF065F46)
                        JournalTokenColor.Screen -> if (isDark) androidx.compose.ui.graphics.Color(0xFFBFDBFE) else androidx.compose.ui.graphics.Color(0xFF1E40AF)
                        JournalTokenColor.Score -> if (isDark) androidx.compose.ui.graphics.Color(0xFFFCD34D) else androidx.compose.ui.graphics.Color(0xFF9A3412)
                        JournalTokenColor.Note -> if (isDark) androidx.compose.ui.graphics.Color(0xFFE6EDF9) else androidx.compose.ui.graphics.Color(0xFF374151)
                    },
                    fontWeight = when (token.color) {
                        JournalTokenColor.Primary, JournalTokenColor.Note -> null
                        else -> FontWeight.SemiBold
                    },
                )
            ) {
                append(token.text)
            }
        }
    }
    Text(text = annotated, style = style, modifier = modifier)
}

private fun journalSummaryTokens(summary: String): List<JournalToken> {
    parseScoreSummary(summary)?.let { (value, game, context) ->
        return listOf(
            JournalToken("Score: ", JournalTokenColor.Primary),
            JournalToken(value, JournalTokenColor.Score),
            JournalToken(" • ", JournalTokenColor.Primary),
            JournalToken(game, JournalTokenColor.Game),
            JournalToken(" ($context)", JournalTokenColor.Screen),
        )
    }

    parseStructuredPracticeSummary(summary)?.let { return it }
    parseStructuredStudySummary(summary)?.let { return it }
    parseStructuredGameNoteSummary(summary)?.let { return it }

    parseBulletGameSummary(summary)?.let { (prefix, game) ->
        return listOf(
            JournalToken(prefix, JournalTokenColor.Primary),
            JournalToken("\n• ", JournalTokenColor.Screen),
            JournalToken(game, JournalTokenColor.Game),
        )
    }

    parseLibrarySummary(summary)?.let { return it }
    parsePracticePlayfieldSummary(summary)?.let { return it }
    parsePracticeRulesheetSummary(summary)?.let { return it }
    parsePracticeVideoSummary(summary)?.let { return it }
    parsePracticeProgressSummary(summary)?.let { return it }
    parsePracticeBrowsedSummary(summary)?.let { return it }

    return listOf(JournalToken(summary, JournalTokenColor.Primary))
}

private fun parseScoreSummary(summary: String): Triple<String, String, String>? {
    if (!summary.startsWith("Score: ")) return null
    val rest = summary.removePrefix("Score: ")
    val bullet = rest.indexOf(" • ")
    val contextStart = rest.lastIndexOf(" (")
    if (bullet <= 0 || contextStart <= bullet || !rest.endsWith(")")) return null
    val value = rest.substring(0, bullet)
    val game = rest.substring(bullet + 3, contextStart).trim()
    val context = rest.substring(contextStart + 2, rest.length - 1)
    if (value.isBlank() || game.isBlank() || context.isBlank()) return null
    return Triple(value, game, context)
}

private fun parseBulletGameSummary(summary: String): Pair<String, String>? {
    val split = summary.split("\n• ", limit = 2)
    if (split.size != 2 || split[1].isBlank()) return null
    return split[0] to split[1]
}

private fun parseStructuredPracticeSummary(summary: String): List<JournalToken>? {
    if (!summary.startsWith("Practice:\n")) return null
    val split = summary.removePrefix("Practice:\n").split("\n• ", limit = 2)
    if (split.size != 2 || split[1].isBlank()) return null
    val bodyLines = split[0].lines().map { it.trimEnd() }.dropLastWhile { it.isBlank() }
    val valueLine = bodyLines.firstOrNull()?.trim().orEmpty().ifBlank { "Practice session" }
    val noteText = bodyLines.drop(1).joinToString("\n").trim().ifBlank { null }
    return buildList {
        add(JournalToken("Practice", JournalTokenColor.Screen))
        add(JournalToken(":\n", JournalTokenColor.Primary))
        add(JournalToken(valueLine, JournalTokenColor.Screen))
        if (noteText != null) {
            add(JournalToken("\n", JournalTokenColor.Primary))
            add(JournalToken(noteText, JournalTokenColor.Note))
        }
        add(JournalToken("\n• ", JournalTokenColor.Screen))
        add(JournalToken(split[1], JournalTokenColor.Game))
    }
}

private fun parseStructuredGameNoteSummary(summary: String): List<JournalToken>? {
    if (!summary.startsWith("Game Note:\n")) return null
    val split = summary.removePrefix("Game Note:\n").split("\n• ", limit = 2)
    if (split.size != 2 || split[1].isBlank()) return null
    return listOf(
        JournalToken("Game Note", JournalTokenColor.Screen),
        JournalToken(":\n", JournalTokenColor.Primary),
        JournalToken(split[0], JournalTokenColor.Note),
        JournalToken("\n• ", JournalTokenColor.Screen),
        JournalToken(split[1], JournalTokenColor.Game),
    )
}

private fun parseStructuredStudySummary(summary: String): List<JournalToken>? {
    val headers = listOf(
        "Rulesheet" to "Rulesheet",
        "Tutorial Video" to "Tutorial Video",
        "Gameplay Video" to "Gameplay Video",
        "Playfield" to "Playfield",
    )
    val header = headers.firstOrNull { summary.startsWith("${it.first}:\n") } ?: return null
    val split = summary.removePrefix("${header.first}:\n").split("\n• ", limit = 2)
    if (split.size != 2 || split[1].isBlank()) return null
    val bodyLines = split[0].lines().map { it.trimEnd() }.dropLastWhile { it.isBlank() }
    val valueLine = bodyLines.firstOrNull()?.trim().orEmpty()
    if (valueLine.isBlank()) return null
    val noteText = bodyLines.drop(1).joinToString("\n").trim().ifBlank { null }
    return buildList {
        add(JournalToken(header.second, JournalTokenColor.Screen))
        add(JournalToken(":\n", JournalTokenColor.Primary))
        when {
            valueLine.startsWith("Progress: ") -> {
                add(JournalToken("Progress", JournalTokenColor.Screen))
                add(JournalToken(": ", JournalTokenColor.Primary))
                add(JournalToken(valueLine.removePrefix("Progress: "), JournalTokenColor.Screen))
            }
            valueLine.equals("Viewed playfield", ignoreCase = true) -> {
                add(JournalToken("Viewed ", JournalTokenColor.Primary))
                add(JournalToken("playfield", JournalTokenColor.Screen))
            }
            else -> add(JournalToken(valueLine, JournalTokenColor.Screen))
        }
        if (noteText != null) {
            add(JournalToken("\n", JournalTokenColor.Primary))
            add(JournalToken(noteText, JournalTokenColor.Note))
        }
        add(JournalToken("\n• ", JournalTokenColor.Screen))
        add(JournalToken(split[1], JournalTokenColor.Game))
    }
}

private fun parseLibrarySummary(summary: String): List<JournalToken>? {
    if (summary.startsWith("Browsed ") && summary.endsWith(" in Library")) {
        val game = summary.removePrefix("Browsed ").removeSuffix(" in Library")
        return listOf(
            JournalToken("Browsed ", JournalTokenColor.Primary),
            JournalToken(game, JournalTokenColor.Game),
            JournalToken(" in ", JournalTokenColor.Primary),
            JournalToken("Library", JournalTokenColor.Screen),
        )
    }
    if (summary.startsWith("Opened ") && summary.endsWith(" rulesheet from Library")) {
        val game = summary.removePrefix("Opened ").removeSuffix(" rulesheet from Library")
        return listOf(
            JournalToken("Opened ", JournalTokenColor.Primary),
            JournalToken(game, JournalTokenColor.Game),
            JournalToken(" rulesheet", JournalTokenColor.Screen),
            JournalToken(" from ", JournalTokenColor.Primary),
            JournalToken("Library", JournalTokenColor.Screen),
        )
    }
    if (summary.startsWith("Opened ") && summary.endsWith(" playfield image from Library")) {
        val game = summary.removePrefix("Opened ").removeSuffix(" playfield image from Library")
        return listOf(
            JournalToken("Opened ", JournalTokenColor.Primary),
            JournalToken(game, JournalTokenColor.Game),
            JournalToken(" playfield image", JournalTokenColor.Screen),
            JournalToken(" from ", JournalTokenColor.Primary),
            JournalToken("Library", JournalTokenColor.Screen),
        )
    }
    if (summary.startsWith("Opened ") && summary.endsWith(" in Library")) {
        val marker = " video for "
        val idx = summary.indexOf(marker)
        if (idx > "Opened ".length) {
            val detail = summary.substring("Opened ".length, idx)
            val game = summary.substring(idx + marker.length, summary.length - " in Library".length)
            return listOf(
                JournalToken("Opened ", JournalTokenColor.Primary),
                JournalToken(detail, JournalTokenColor.Screen),
                JournalToken(" video", JournalTokenColor.Screen),
                JournalToken(" for ", JournalTokenColor.Primary),
                JournalToken(game, JournalTokenColor.Game),
                JournalToken(" in ", JournalTokenColor.Primary),
                JournalToken("Library", JournalTokenColor.Screen),
            )
        }
    }
    return null
}

private fun parsePracticePlayfieldSummary(summary: String): List<JournalToken>? {
    if (!summary.startsWith("Viewed ")) return null
    val marker = " playfield"
    val idx = summary.indexOf(marker)
    if (idx <= "Viewed ".length) return null
    val game = summary.substring("Viewed ".length, idx)
    if (game.isBlank()) return null
    val suffix = summary.substring(idx + marker.length)
    val note = when {
        suffix.isEmpty() -> null
        suffix.startsWith(": ") -> suffix.removePrefix(": ")
        else -> return null
    }
    return buildList {
        add(JournalToken("Viewed ", JournalTokenColor.Primary))
        add(JournalToken(game, JournalTokenColor.Game))
        add(JournalToken(" ", JournalTokenColor.Primary))
        add(JournalToken("playfield", JournalTokenColor.Screen))
        if (!note.isNullOrEmpty()) add(JournalToken(": $note", JournalTokenColor.Primary))
    }
}

private fun parsePracticeRulesheetSummary(summary: String): List<JournalToken>? {
    if (!summary.startsWith("Read ") || !summary.endsWith(" rulesheet")) return null
    val body = summary.removePrefix("Read ").removeSuffix(" rulesheet")
    val ofIdx = body.indexOf(" of ")
    return if (ofIdx > 0) {
        val progress = body.substring(0, ofIdx)
        val game = body.substring(ofIdx + 4)
        if (progress.isBlank() || game.isBlank()) null else listOf(
            JournalToken("Read ", JournalTokenColor.Primary),
            JournalToken(progress, JournalTokenColor.Screen),
            JournalToken(" of ", JournalTokenColor.Primary),
            JournalToken(game, JournalTokenColor.Game),
            JournalToken(" ", JournalTokenColor.Primary),
            JournalToken("rulesheet", JournalTokenColor.Screen),
        )
    } else {
        if (body.isBlank()) null else listOf(
            JournalToken("Read ", JournalTokenColor.Primary),
            JournalToken(body, JournalTokenColor.Game),
            JournalToken(" ", JournalTokenColor.Primary),
            JournalToken("rulesheet", JournalTokenColor.Screen),
        )
    }
}

private fun parsePracticeVideoSummary(summary: String): List<JournalToken>? {
    if (summary.startsWith("Tutorial for ")) {
        val idx = summary.indexOf(": ")
        if (idx > "Tutorial for ".length) {
            val game = summary.substring("Tutorial for ".length, idx)
            val value = summary.substring(idx + 2)
            if (game.isNotBlank()) {
                return listOf(
                    JournalToken("Tutorial", JournalTokenColor.Screen),
                    JournalToken(" for ", JournalTokenColor.Primary),
                    JournalToken(game, JournalTokenColor.Game),
                    JournalToken(": ", JournalTokenColor.Primary),
                    JournalToken(value, JournalTokenColor.Primary),
                )
            }
        }
    }
    if (summary.startsWith("Gameplay for ")) {
        val idx = summary.indexOf(": ")
        if (idx > "Gameplay for ".length) {
            val game = summary.substring("Gameplay for ".length, idx)
            val value = summary.substring(idx + 2)
            if (game.isNotBlank()) {
                return listOf(
                    JournalToken("Gameplay", JournalTokenColor.Screen),
                    JournalToken(" for ", JournalTokenColor.Primary),
                    JournalToken(game, JournalTokenColor.Game),
                    JournalToken(": ", JournalTokenColor.Primary),
                    JournalToken(value, JournalTokenColor.Primary),
                )
            }
        }
    }
    if (summary.startsWith("Updated tutorial progress for ")) {
        val game = summary.removePrefix("Updated tutorial progress for ")
        if (game.isNotBlank()) {
            return listOf(
                JournalToken("Updated ", JournalTokenColor.Primary),
                JournalToken("tutorial", JournalTokenColor.Screen),
                JournalToken(" progress for ", JournalTokenColor.Primary),
                JournalToken(game, JournalTokenColor.Game),
            )
        }
    }
    if (summary.startsWith("Updated gameplay progress for ")) {
        val game = summary.removePrefix("Updated gameplay progress for ")
        if (game.isNotBlank()) {
            return listOf(
                JournalToken("Updated ", JournalTokenColor.Primary),
                JournalToken("gameplay", JournalTokenColor.Screen),
                JournalToken(" progress for ", JournalTokenColor.Primary),
                JournalToken(game, JournalTokenColor.Game),
            )
        }
    }
    return null
}

private fun parsePracticeProgressSummary(summary: String): List<JournalToken>? {
    if (summary.startsWith("Practice progress ")) {
        val onIdx = summary.indexOf(" on ")
        if (onIdx > "Practice progress ".length) {
            val value = summary.substring("Practice progress ".length, onIdx)
            val game = summary.substring(onIdx + 4)
            if (value.isNotBlank() && game.isNotBlank()) {
                return listOf(
                    JournalToken("Practice", JournalTokenColor.Screen),
                    JournalToken(" progress ", JournalTokenColor.Primary),
                    JournalToken(value, JournalTokenColor.Screen),
                    JournalToken(" on ", JournalTokenColor.Primary),
                    JournalToken(game, JournalTokenColor.Game),
                )
            }
        }
    }
    if (summary.startsWith("Logged practice for ")) {
        val game = summary.removePrefix("Logged practice for ")
        if (game.isNotBlank()) {
            return listOf(
                JournalToken("Logged ", JournalTokenColor.Primary),
                JournalToken("practice", JournalTokenColor.Screen),
                JournalToken(" for ", JournalTokenColor.Primary),
                JournalToken(game, JournalTokenColor.Game),
            )
        }
    }
    return null
}

private fun parsePracticeBrowsedSummary(summary: String): List<JournalToken>? {
    if (!summary.startsWith("Browsed ") || summary.endsWith(" in Library")) return null
    val game = summary.removePrefix("Browsed ")
    if (game.isBlank()) return null
    return listOf(
        JournalToken("Browsed ", JournalTokenColor.Primary),
        JournalToken(game, JournalTokenColor.Game),
    )
}
